LWCE_Core/Src/XComLongWarCommunityEdition/Classes/LWCETacticalVisibilityHelper.uc

```diff
@@ -254,22 +282,140 @@ protected function InitializeHelpers()
+    class'XComWorldData'.static.GetWorldData().SetTileBlockedByUnitFlag(m_kCoverUsingHelper);
+
+    ClearVisHelperTileClaims(/* bForce */ true);
+}
+
+/// <summary>
+/// Checks if the game is in a state where updating visibility data (such as by moving the helper unit, or
+/// updating their cover states) is going to break something.
+/// </summary>
+protected function bool CanSafelyUpdateVisibility()
+{
+    local XGAction kAction;
+    local XGAction_Targeting kTargetingAction;
+    local XGUnit kActiveUnit;
+
+    if (!m_bInitialized)
+    {
+        return false;
+    }
+
+    kActiveUnit = TacticalController().m_kActiveUnit;
+
+    if (kActiveUnit == none)
+    {
+        return true;
+    }
+
+    kAction = kActiveUnit.GetAction();
+
+    if (kAction == none)
+    {
+        return true;
+    }
+
+    kTargetingAction = XGAction_Targeting(kActiveUnit.GetAction());
+
+    if (kTargetingAction != none && kTargetingAction.m_kShot != none && kTargetingAction.m_kShot.IsA('XGAbility_Grapple'))
+    {
+        return true;
+    }
+
+    // XGAction_Path is used when choosing the unit's destination; it's the only action other than targeting a grapple
+    // that we can safely update visibility during. IsPerformingAction does not count it, hence the separate check.
+    if (kAction.IsA('XGAction_Path'))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+/// <summary>
+/// Marks tiles occupied by vis helpers as unoccupied, unless there is an actual unit occupying the same tile.
+/// </summary>
+/// <param name="bForce">If true, tile claims will be modified even if it's not currently safe to do so.</param>
+protected function ClearVisHelperTileClaims(optional bool bForce = false)
+{
+    local bool bClearCoverHelper, bClearNonCoverHelper;
+    local int iCoverHelperX, iCoverHelperY, iCoverHelperZ;
+    local int iNonCoverHelperX, iNonCoverHelperY, iNonCoverHelperZ;
+    local int iUnitX, iUnitY, iUnitZ;
+    local XComWorldData kWorldData;
+    local XGUnit kUnit;
+
+    if (!bForce && !CanSafelyUpdateVisibility())
+    {
+        return;
+    }
+
+    bClearCoverHelper = true;
+    bClearNonCoverHelper = true;
+    kWorldData = class'XComWorldData'.static.GetWorldData();
+
+    kWorldData.GetTileCoordinatesFromPosition(m_kCoverUsingHelper.Location, iCoverHelperX, iCoverHelperY, iCoverHelperZ);
+    kWorldData.GetTileCoordinatesFromPosition(m_kNonCoverUsingHelper.Location, iNonCoverHelperX, iNonCoverHelperY, iNonCoverHelperZ);
+
+    foreach AllActors(class'XGUnit', kUnit)
+    {
+        if (IsVisHelper(kUnit))
+        {
+            continue;
+        }
+
+        kWorldData.GetTileCoordinatesFromPosition(kUnit.Location, iUnitX, iUnitY, iUnitZ);
+
+        // Our vis helpers don't always quite match unit positions on the Z axis, so add a small tolerance in that dimension
+        if (iUnitX == iCoverHelperX && iUnitY == iCoverHelperY && Abs(iUnitZ - iCoverHelperZ) <= 2)
```

Comment:
Should we log a warning in these cases? As far as I understand we never intentionally move the helper to the same tile as another unit.

Also: It's so weird that even the tile coordinates are different, I thought you just meant the unit locations :open_mouth: